networkd/network_server.ml

```diff
@@ -474,12 +474,96 @@ module Bridge = struct
 			| Bridge -> Sysfs.get_all_bridges ()
 		) ()
 
+	let destroy _ dbg ?(force=false) ~name () =
+		Debug.with_thread_associated dbg (fun () ->
+			Interface.bring_down () dbg ~name;
+			match !backend_kind with
+			| Openvswitch ->
+				if Ovs.get_vlans name = [] || force then begin
+					debug "Destroying bridge %s" name;
+					remove_config name;
+					List.iter (fun dev ->
+						Interface.set_ipv4_conf () dbg ~name:dev ~conf:None4;
+						Interface.bring_down () dbg ~name:dev
+					) (Ovs.bridge_to_interfaces name);
+					Interface.set_ipv4_conf () dbg ~name ~conf:None4;
+					ignore (Ovs.destroy_bridge name)
+				end else
+					debug "Not destroying bridge %s, because it has VLANs on top" name
+			| Bridge ->
+				let ifs = Sysfs.bridge_to_interfaces name in
+				let vlans_on_this_parent =
+					let interfaces = List.filter (fun n ->
+						String.startswith "eth" n || String.startswith "bond" n
+					) ifs in
+					match interfaces with
+					| [] -> []
+					| interface :: _ ->
+						List.filter (String.startswith (interface ^ ".")) (Sysfs.list ())
+				in
+				if vlans_on_this_parent = [] || force then begin
+					debug "Destroying bridge %s" name;
+					remove_config name;
+					List.iter (fun dev ->
+						Interface.set_ipv4_conf () dbg ~name:dev ~conf:None4;
+						Brctl.destroy_port name dev;
+						Interface.bring_down () dbg ~name:dev;
+						if Linux_bonding.is_bond_device dev then
+							Linux_bonding.remove_bond_master dev;
+							if (String.startswith "eth" dev || String.startswith "bond" dev) && String.contains dev '.' then begin
+								ignore (Ip.destroy_vlan dev);
+								let n = String.length dev in
+								if String.sub dev (n - 2) 2 = ".0" && need_enic_workaround () then
+									let vlan_base = String.sub dev 0 (n - 2) in
+									if Linux_bonding.is_bond_device vlan_base then
+										Linux_bonding.remove_bond_master (String.sub dev 0 (n - 2))
+							end;
+					) ifs;
+					Interface.set_ipv4_conf () dbg ~name ~conf:None4;
+					ignore (Brctl.destroy_bridge name)
+				end else
+					debug "Not destroying bridge %s, because it has VLANs on top" name
+	) ()
+
+	let destroy_existing_vlan_bridge dbg ~vlan ~parent () =
+		let existing_bridge =
+			match !backend_kind with
+			| Openvswitch ->
+				let bridges = Ovs.list_bridges () in
```

Comment:
This may be quite inefficient in case there are large numbers of VLANs, especially because the `Ovs` call below means forking ovs-vsctl for each bridge. There must be a more sophisticated query we can do.